flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepointTest.java

```diff
@@ -0,0 +1,248 @@
+            this.state = state;
+            this.executionGraphTerminationFuture = executionGraphTerminationFuture;
+            this.operationCompletionFuture = operationCompletionFuture;
+            this.savepointFuture = savepointFuture;
+        }
+
+        public StopWithSavepoint getState() {
+            return state;
+        }
+
+        public CompletableFuture<JobStatus> getExecutionGraphTerminationFuture() {
+            return executionGraphTerminationFuture;
+        }
+
+        public CompletableFuture<String> getOperationCompletionFuture() {
+            return operationCompletionFuture;
+        }
+
+        public CompletableFuture<CompletedCheckpoint> getSavepointFuture() {
+            return savepointFuture;
+        }
+    }
+
+    private StopWithSavepointEnvironment createStopWithSavepointEnvironment(
+            MockExecutingStateWithFailureHandlerContext ctx) throws IOException {
+        CompletableFuture<JobStatus> executionGraphTerminationFuture = new CompletableFuture<>();
+        CompletableFuture<String> operationCompletionFuture = new CompletableFuture<>();
+        CompletableFuture<CompletedCheckpoint> savepointFuture = new CompletableFuture<>();
+        final ExecutionGraph executionGraph =
+                new MockExecutionGraph(executionGraphTerminationFuture);
+        executionGraph.transitionToRunning();
+        final ExecutionGraphHandler executionGraphHandler =
+                new ExecutionGraphHandler(
+                        executionGraph,
+                        log,
+                        ctx.getMainThreadExecutor(),
+                        ctx.getMainThreadExecutor());
+        OperatorCoordinatorHandler operatorCoordinatorHandler =
+                new OperatorCoordinatorHandler(
+                        executionGraph,
+                        (throwable) -> {
+                            throw new RuntimeException("Error in test", throwable);
+                        });
+
+        CheckpointCoordinator checkpointCoordinator =
+                new MockCheckpointCoordinator(savepointFuture);
+
+        StopWithSavepoint stopWithSavepoint =
+                new StopWithSavepoint(
+                        ctx,
+                        executionGraph,
+                        executionGraphHandler,
+                        operatorCoordinatorHandler,
+                        log,
+                        ClassLoader.getSystemClassLoader(),
+                        checkpointCoordinator,
+                        "savepointTargetDir",
+                        false,
+                        operationCompletionFuture);
+        return new StopWithSavepointEnvironment(
+                stopWithSavepoint,
+                executionGraphTerminationFuture,
+                operationCompletionFuture,
+                savepointFuture);
+    }
+
+    private static class MockCheckpointCoordinator extends CheckpointCoordinator {
```

Comment:
Let's not continue with this pattern. Extending concrete classes is a recipe for a lot of rabbit holes.

Maybe an easier solution could be to define the interface the `StopWithSavepoint` state requires and then have some adapters to map it to the underlying `CheckpointCoordinator` (or even let the `CheckpointCoordinator` implement this interface).